File: src/iterator.cc
#include "iterator.h"
#include "marker-index.h"

Iterator::Iterator(MarkerIndex *marker_index) :
  marker_index {marker_index},
  node {nullptr} {}

void Iterator::Reset() {
  node = marker_index->root;
  if (node) {
    node_offset = node->left_extent;
  }
  left_ancestor_offset = Point(0, 0);
  right_ancestor_offset = Point::Max();
  left_ancestor_offset_stack.clear();
  right_ancestor_offset_stack.clear();
}

Node* Iterator::InsertMarkerStart(const MarkerId &id, const Point &start_offset, const Point &end_offset) {
  Reset();

  if (!node) {
    return marker_index->root = new Node(nullptr, start_offset);
  }

  while (true) {
    switch (start_offset.Compare(node_offset)) {
      case 0:
        MarkRight(id, start_offset, end_offset);
        return node;
      case -1:
        MarkRight(id, start_offset, end_offset);
        if (node->left) {
          DescendLeft();
          break;
        } else {
          InsertLeftChild(start_offset);
          DescendLeft();
          MarkRight(id, start_offset, end_offset);
          return node;
        }
      case 1:
        if (node->right) {
          DescendRight();
          break;
        } else {
          InsertRightChild(start_offset);
          DescendRight();
          MarkRight(id, start_offset, end_offset);
          return node;
        }
    }
  }
}

Node* Iterator::InsertMarkerEnd(const MarkerId &id, const Point &start_offset, const Point &end_offset) {
  Reset();

  if (!node) {
    return marker_index->root = new Node(nullptr, end_offset);
  }

  while (true) {
    switch (end_offset.Compare(node_offset)) {
      case 0:
        MarkLeft(id, start_offset, end_offset);
        return node;
      case -1:
        if (node->left) {
          DescendLeft();
          break;
        } else {
          InsertLeftChild(end_offset);
          DescendLeft();
          MarkLeft(id, start_offset, end_offset);
          return node;
        }
      case 1:
        MarkLeft(id, start_offset, end_offset);
        if (node->right) {
          DescendRight();
          break;
        } else {
          InsertRightChild(end_offset);
          DescendRight();
          MarkLeft(id, start_offset, end_offset);
          return node;
        }
    }
  }
}

void Iterator::DescendLeft() {
  left_ancestor_offset_stack.push_back(left_ancestor_offset);
  right_ancestor_offset_stack.push_back(right_ancestor_offset);

  right_ancestor_offset = node_offset;
  node = node->left;
  node_offset = left_ancestor_offset.Traverse(node->left_extent);
}

void Iterator::DescendRight() {
  left_ancestor_offset_stack.push_back(left_ancestor_offset);
  right_ancestor_offset_stack.push_back(right_ancestor_offset);

  left_ancestor_offset = node_offset;
  node = node->right;
  node_offset = left_ancestor_offset.Traverse(node->left_extent);
}

void Iterator::MarkRight(const MarkerId &id, const Point &start_offset, const Point &end_offset) {
  if (left_ancestor_offset < start_offset
    && start_offset <= node_offset
    && right_ancestor_offset <= end_offset) {
    node->right_marker_ids.insert(id);
  }
}

void Iterator::MarkLeft(const MarkerId &id, const Point &start_offset, const Point &end_offset) {
  if (!node_offset.IsZero() && start_offset <= left_ancestor_offset && node_offset <= end_offset) {
    node->left_marker_ids.insert(id);
  }
}

Node* Iterator::InsertLeftChild(const Point &offset) {
  return node->left = new Node(node, offset.Traversal(left_ancestor_offset));
}

Node* Iterator::InsertRightChild(const Point &offset) {
  return node->right = new Node(node, offset.Traversal(node_offset));
}

File: src/js/iterator.js
import Node from './node'
import {addToSet} from './helpers'
import {compare, isZero, traversal, traverse} from './point-helpers'

export default class Iterator {
  constructor (markerIndex) {
    this.markerIndex = markerIndex
  }

  reset () {
    this.node = this.markerIndex.root
    this.nodePosition = this.node ? this.node.leftExtent : null
    this.leftAncestorPosition = {row: 0, column: 0}
    this.rightAncestorPosition = {row: Infinity, column: Infinity}
    this.leftAncestorPositionStack = []
    this.rightAncestorPositionStack = []
  }

  insertMarkerStart (markerId, startPosition, endPosition) {
    this.reset()

    if (!this.node) {
      let node = new Node(null, startPosition)
      this.markerIndex.root = node
      return node
    }

    while (true) {
      let comparison = compare(startPosition, this.nodePosition)
      if (comparison === 0) {
        this.markRight(markerId, startPosition, endPosition)
        return this.node
      } else if (comparison < 0) {
        this.markRight(markerId, startPosition, endPosition)
        if (this.node.left) {
          this.descendLeft()
        } else {
          this.insertLeftChild(startPosition)
          this.descendLeft()
          this.markRight(markerId, startPosition, endPosition)
          return this.node
        }
      } else { // startPosition > this.nodePosition
        if (this.node.right) {
          this.descendRight()
        } else {
          this.insertRightChild(startPosition)
          this.descendRight()
          this.markRight(markerId, startPosition, endPosition)
          return this.node
        }
      }
    }
  }

  insertMarkerEnd (markerId, startPosition, endPosition) {
    this.reset()

    if (!this.node) {
      let node = new Node(null, endPosition)
      this.markerIndex.root = node
      return node
    }

    while (true) {
      let comparison = compare(endPosition, this.nodePosition)
      if (comparison === 0) {
        this.markLeft(markerId, startPosition, endPosition)
        return this.node
      } else if (comparison < 0) {
        if (this.node.left) {
          this.descendLeft()
        } else {
          this.insertLeftChild(endPosition)
          this.descendLeft()
          this.markLeft(markerId, startPosition, endPosition)
          return this.node
        }
      } else { // endPosition > this.nodePosition
        this.markLeft(markerId, startPosition, endPosition)
        if (this.node.right) {
          this.descendRight()
        } else {
          this.insertRightChild(endPosition)
          this.descendRight()
          this.markLeft(markerId, startPosition, endPosition)
          return this.node
        }
      }
    }
  }

  insertSpliceBoundary (position, isInsertionEnd) {
    this.reset()

    while (true) {
      let comparison = compare(position, this.nodePosition)
      if (comparison === 0 && !isInsertionEnd) {
        return this.node
      } else if (comparison < 0) {
        if (this.node.left) {
          this.descendLeft()
        } else {
          this.insertLeftChild(position)
          return this.node.left
        }
      } else { // position > this.nodePosition
        if (this.node.right) {
          this.descendRight()
        } else {
          this.insertRightChild(position)
          return this.node.right
        }
      }
    }
  }

  findIntersecting (start, end, resultSet) {
    this.reset()
    if (!this.node) return

    while (true) {
      if (compare(start, this.nodePosition) < 0) {
        if (this.node.left) {
          this.checkIntersection(start, end, resultSet)
          this.descendLeft()
        } else {
          break
        }
      } else {
        if (this.node.right) {
          this.checkIntersection(start, end, resultSet)
          this.descendRight()
        } else {
          break
        }
      }
    }

    do {
      this.checkIntersection(start, end, resultSet)
      this.moveToSuccessor()
    } while (this.node && compare(this.nodePosition, end) <= 0)
  }

  findContaining (position, resultSet) {
    this.reset()
    if (!this.node) return

    while (true) {
      this.checkIntersection(position, position, resultSet)

      if (compare(position, this.nodePosition) < 0) {
        if (this.node.left) {
          this.descendLeft()
        } else {
          break
        }
      } else {
        if (this.node.right) {
          this.descendRight()
        } else {
          break
        }
      }
    }
  }

  findContainedIn (start, end, resultSet) {
    this.reset()
    if (!this.node) return

    this.seekToFirstNodeGreaterThanOrEqualTo(start)

    let started = new Set()
    while (this.node && compare(this.nodePosition, end) <= 0) {
      addToSet(started, this.node.startMarkerIds)
      this.node.endMarkerIds.forEach(function (markerId) {
        if (started.has(markerId)) {
          resultSet.add(markerId)
        }
      })
      this.moveToSuccessor()
    }
  }

  findStartingIn (start, end, resultSet) {
    this.reset()
    if (!this.node) return

    this.seekToFirstNodeGreaterThanOrEqualTo(start)

    while (this.node && compare(this.nodePosition, end) <= 0) {
      addToSet(resultSet, this.node.startMarkerIds)
      this.moveToSuccessor()
    }
  }

  findEndingIn (start, end, resultSet) {
    this.reset()
    if (!this.node) return

    this.seekToFirstNodeGreaterThanOrEqualTo(start)

    while (this.node && compare(this.nodePosition, end) <= 0) {
      addToSet(resultSet, this.node.endMarkerIds)
      this.moveToSuccessor()
    }
  }

  dump () {
    this.reset()

    while (this.node && this.node.left) this.descendLeft()

    let snapshot = {}

    while (this.node) {
      this.node.startMarkerIds.forEach(markerId => {
        snapshot[markerId] = {start: this.nodePosition, end: null}
      })

      this.node.endMarkerIds.forEach(markerId => {
        snapshot[markerId].end = this.nodePosition
      })

      this.moveToSuccessor()
    }

    return snapshot
  }

  seekToFirstNodeGreaterThanOrEqualTo (position) {
    while (true) {
      let comparison = compare(position, this.nodePosition)

      if (comparison === 0) {
        break
      } else if (comparison < 0) {
        if (this.node.left) {
          this.descendLeft()
        } else {
          break
        }
      } else {
        if (this.node.right) {
          this.descendRight()
        } else {
          break
        }
      }
    }

    if (compare(this.nodePosition, position) < 0) this.moveToSuccessor()
  }

  markLeft (markerId, startPosition, endPosition) {
    if (!isZero(this.nodePosition) && compare(startPosition, this.leftAncestorPosition) <= 0 && compare(this.nodePosition, endPosition) <= 0) {
      this.node.leftMarkerIds.add(markerId)
    }
  }

  markRight (markerId, startPosition, endPosition) {
    if (compare(this.leftAncestorPosition, startPosition) < 0 &&
      compare(startPosition, this.nodePosition) <= 0 &&
      compare(this.rightAncestorPosition, endPosition) <= 0) {
      this.node.rightMarkerIds.add(markerId)
    }
  }

  ascend () {
    if (this.node.parent) {
      if (this.node.parent.left === this.node) {
        this.nodePosition = this.rightAncestorPosition
      } else {
        this.nodePosition = this.leftAncestorPosition
      }
      this.leftAncestorPosition = this.leftAncestorPositionStack.pop()
      this.rightAncestorPosition = this.rightAncestorPositionStack.pop()
      this.node = this.node.parent
    } else {
      this.node = null
      this.nodePosition = null
      this.leftAncestorPosition = {row: 0, column: 0}
      this.rightAncestorPosition = {row: Infinity, column: Infinity}
    }
  }

  descendLeft () {
    this.leftAncestorPositionStack.push(this.leftAncestorPosition)
    this.rightAncestorPositionStack.push(this.rightAncestorPosition)

    this.rightAncestorPosition = this.nodePosition
    this.node = this.node.left
    this.nodePosition = traverse(this.leftAncestorPosition, this.node.leftExtent)
  }

  descendRight () {
    this.leftAncestorPositionStack.push(this.leftAncestorPosition)
    this.rightAncestorPositionStack.push(this.rightAncestorPosition)

    this.leftAncestorPosition = this.nodePosition
    this.node = this.node.right
    this.nodePosition = traverse(this.leftAncestorPosition, this.node.leftExtent)
  }

  moveToSuccessor () {
    if (!this.node) return

    if (this.node.right) {
      this.descendRight()
      while (this.node.left) {
        this.descendLeft()
      }
    } else {
      while (this.node.parent && this.node.parent.right === this.node) {
        this.ascend()
      }
      this.ascend()
    }
  }

  insertLeftChild (position) {
    this.node.left = new Node(this.node, traversal(position, this.leftAncestorPosition))
  }

  insertRightChild (position) {
    this.node.right = new Node(this.node, traversal(position, this.nodePosition))
  }

  checkIntersection (start, end, resultSet) {
    if (compare(this.leftAncestorPosition, end) <= 0 && compare(start, this.nodePosition) <= 0) {
      addToSet(resultSet, this.node.leftMarkerIds)
    }

    if (compare(start, this.nodePosition) <= 0 && compare(this.nodePosition, end) <= 0) {
      addToSet(resultSet, this.node.startMarkerIds)
      addToSet(resultSet, this.node.endMarkerIds)
    }

    if (compare(this.nodePosition, end) <= 0 && compare(start, this.rightAncestorPosition) <= 0) {
      addToSet(resultSet, this.node.rightMarkerIds)
    }
  }
}
